src/services/npdi.ts

```diff
@@ -1,24 +1,32 @@
-import path from 'node:path';
 import express from 'express';
-import { fileErrCallback } from '@/util';
-import { __appRoot } from '@/app-root';
 import { restrictHostnames } from '@/middleware/host-limit';
 import { config } from '@/config-manager';
+import { getCdnFileAsStream, streamFileToResponse } from '@/cdn';
+import { getTaskFileByDataID } from '@/database';
 
 const npdi = express.Router();
 
-npdi.get('/p01/data/1/:titleHash/:dataID/:fileHash', (request, response) => {
-	const { titleHash, fileHash } = request.params;
-	const contentPath = path.normalize(`${__appRoot}/../cdn/content/encrypted/${titleHash}/${fileHash}`);
+npdi.get('/p01/data/1/:bossAppId/:dataId/:fileHash', async (request, response) => {
+	const { dataId, fileHash, bossAppId } = request.params;
 
-	response.sendFile(contentPath, {
-		headers: {
-			// * The misspelling here is intentional, it's what the official server sets
-			'Content-Type': 'applicatoin/octet-stream',
-			'Content-Disposition': 'attachment',
-			'Content-Transfer-Encoding': 'binary'
-		}
-	}, fileErrCallback(response));
+	const file = await getTaskFileByDataID(BigInt(dataId));
+	if (!file) {
+		return response.sendStatus(404);
+	}
+	if (file.hash !== fileHash || file.boss_app_id !== bossAppId) {
+		return response.sendStatus(404);
+	}
+
+	const fileStream = await getCdnFileAsStream('taskFile', file.file_key);
+	if (!fileStream) {
+		throw new Error('File not found in CDN');
+	}
+	return streamFileToResponse(response, fileStream, {
```

Comment:
```suggestion
	}

	if (file.hash !== fileHash || file.boss_app_id !== bossAppId) {
		return response.sendStatus(404);
	}

	const fileStream = await getCdnFileAsStream('taskFile', file.file_key);
	if (!fileStream) {
		throw new Error('File not found in CDN');
	}

	return streamFileToResponse(response, fileStream, {
```

Just a small formatting nit for consistency

Reply:
Fixed